hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenAirdropHandler.java

```diff
@@ -0,0 +1,563 @@
+import com.hedera.node.app.service.token.records.TokenAirdropRecordBuilder;
+import com.hedera.node.app.spi.fees.FeeContext;
+import com.hedera.node.app.spi.fees.Fees;
+import com.hedera.node.app.spi.workflows.HandleContext;
+import com.hedera.node.app.spi.workflows.HandleException;
+import com.hedera.node.app.spi.workflows.PreCheckException;
+import com.hedera.node.app.spi.workflows.PreHandleContext;
+import com.hedera.node.app.spi.workflows.TransactionHandler;
+import com.hedera.node.config.data.TokensConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * This class contains all workflow-related functionality regarding {@link
+ * HederaFunctionality#TOKEN_AIRDROP}.
+ */
+@Singleton
+public class TokenAirdropHandler implements TransactionHandler {
+
+    private final CryptoTransferValidator validator;
+
+    /**
+     * Default constructor for injection.
+     */
+    @Inject
+    public TokenAirdropHandler(@NonNull final CryptoTransferValidator validator) {
+        this.validator = validator;
+    }
+
+    @Override
+    public void preHandle(@NonNull final PreHandleContext context) throws PreCheckException {
+        requireNonNull(context);
+        pureChecks(context.body());
+
+        final var op = context.body().tokenAirdropOrThrow();
+        final var accountStore = context.createStore(ReadableAccountStore.class);
+        final var tokenStore = context.createStore(ReadableTokenStore.class);
+
+        for (final var transfers : op.tokenTransfers()) {
+            final var tokenID = transfers.tokenOrThrow();
+            final var tokenMeta = tokenStore.getTokenMeta(transfers.tokenOrElse(TokenID.DEFAULT));
+            validateTruePreCheck(tokenMeta != null, INVALID_TOKEN_ID);
+            checkFungibleTokenTransfers(tokenID, transfers.transfers(), context, accountStore);
+            checkNftTransfers(tokenID, transfers.nftTransfers(), context, tokenMeta, accountStore);
+        }
+    }
+
+    @Override
+    public void pureChecks(@NonNull final TransactionBody txn) throws PreCheckException {
+        requireNonNull(txn);
+        final var op = txn.tokenAirdropOrThrow();
+        validator.airdropsPureChecks(op);
+    }
+
+    @Override
+    public void handle(@NonNull final HandleContext context) throws HandleException {
+        requireNonNull(context);
+        final var txn = context.body();
+        final var op = txn.tokenAirdropOrThrow();
+        final var pendingStore = context.storeFactory().writableStore(WritableAirdropStore.class);
+        final var accountStore = context.storeFactory().writableStore(WritableAccountStore.class);
+        var recordBuilder = context.recordBuilders().getOrCreate(TokenAirdropRecordBuilder.class);
```

Comment:
If the logic looks like this, it will look simple and readable. What do you think ? 
```
final var airdropTransfer = translateToTransfer(airdrop)
if (availableAutoAssociationsSlots) {
    // auto-creation, association, custom fees are already charged when we dispatch. 
    // We can think of not externalizing child records if we want
    final var child = dispatch(airdropTransfer)
    throwIfNotSuccessful(child)
} else {
    chargeCustomFees();
    chargeAssociationFees();
    createPendingAirdrop();
    putIntoState();
}
```